### ISSUE TITLE
Working with Formatted text in the editor

### ISSUE THREAD

**Issue by rchouine** [Mediawiki-wysiwyg/WYSIWYG-CKeditor#45]
When I select the format "Formatted" in the editor, I can arrange my text in the editor itself. 
When I come back, my formatted text is replaced by a special tag, and I lose some of my format in the dialog.
Is there a way to keep the "pre" tags so it is handeled by the editor?

I tried to change the page "\WYSIWYG\CKeditorParser.body.php", function "strip" to add a case:
                    case 'pre':
                        //Keep original data, replace all inner tags to avoid further interpretation
                        $output = '< p r e >' . htmlentities($content) . '< / p r e >';
                        break;

This code works fine for me. so the "pre" tag is not replaced by a fake object.

Is there a better way to manage this without changing the "CkEditor" files?

**Comment by riparap** [comment]
Format "formatted" seems to work like you have described. As far as I know "special" element will not keep indents made by spaces, it would require ackward modification where spaces are converted to something else like "fckSpace" and then back to real spaces little like with syntaxhighlight -tag. If your modification takes care of problem I think it is valid solution.... I will also test it in my wiki.

**Comment by rchouine** [comment]
I had concerns modifying "CkEditor" files thinking you only support "MediaWiki" files.
Let me know when you try my code, or if you want me to apply it in one of your branches.

**Comment by riparap** [comment]
I am not sure, but is it so that your fix does not keep possible attributes of pre -tag, fex:

``` sh
<pre style="color: green; text-decoration: line-through;">
       Text is '''preformatted''' 
                                with a style and 
                          ''markups'' '''''cannot''''' be done
</pre>
```

``` sh
<pre style="color: red; text-decoration: line-through;">
Text is '''preformatted''' 

             with a style and 

                         ''markups'' '''''cannot''''' be done
</pre>
```

I am able to fix above cases with special -element using fckSPACE / ' ' pairs to preserve spaces. Are there other formats than spaces which you have lost with special -element? 

On the other hand, special -element is rather clumsy way to edit text so if we can find better solution for direct editing of this kind of text on page, it would be good, so in that sence your approach is better.

**Comment by riparap** [comment]
I have pushed modified "special" element into temporary test branch "CKeditor_v4_preTag". It now preserves spaces and also keeps possible attributes of pre tag. Does this work with your pages?

Your code is included too, but it is disabled by "wrong" name of tag = "case 'pre-rchouine':". If you activate your modification, my fix does not have any effect, so you can easily test your code too.

it seems to be so that format "Formatted" is not "compatible" with f.ex text color and font selections, because they use span tag and format "formatted" is using pre -tag. If you select font and font color for text and then select format "Formatted", it will override everything else resulting  html code of span -tags on page. Because of this I am voting for using special -element with pre -tags. But I agree, it is not the most user friendly solution.

**Comment by rchouine** [comment]
Thanks, the special dialog is a lot better this way. But I still prefer to see the pre sections directly in the main editor, a lot more visual. So I added the attributes of the tag so they are not lost anymore.
 I also renamed the  "case 'pre-rchouine' to 'pre' so you can test it directly.

I agree that changing the format makes you lose the "pre" tags, buit I suppose that users should know what they are doing :-)

**Comment by riparap** [comment]
In case there were more than one sequeantial different blocks of pre- tags on page, using source button few times corrupted the contents of page. After I added line feed in .php after ending pre- tag, it seems like it now works better. 

I slightly modied it further so that there is now possibility to select if special- element is used with pre-tags or if text is placed on page directly. If $wgFCKEditorSpecialElementWithPreTag = 1 in LocalSettings.php, special element is used in case pre-tag contains attributes. Otherwise pre -tag is always placed on page directly as text.

Blocks with pre -tags should now be displayed on light grey background in similar way as in MW. I hope it works properly, because I am not sure about the proper way of this fix.

These modifications are available under branch "CKeditor_v4_preTag".

I also fixed extra line feed problem with magic tags like '''**TOC**''' (occurred when source button was pressed multible times, each toggling caused one extra line feed to be added after '**TOC**' element).

**Comment by rchouine** [comment]
Sorry for the late feedback.
I just tried the latest version and it seems fine to me. I like the gray background in the edit mode too.
The special objects put the icon inline with the code before and after the tag. I prefer this version.
I don't know what other people think, I hope they'll agree :)

**Comment by riparap** [comment]
Nb... thank you for your feedback. I think as long as modification works, other people will be happy too :-). I think too that having pre -tag directly on page is better than special -element, because special element makes editing of text difficult.

I will anyway wait a few days for more comments before I merge this into main branch.

**Comment by riparap** [comment]
This has been merged into branch CKeditor_V4 at 17.02.2015 (version 1.5.6_0 [B551+10.02.2015]).
